contracts/vortex/CarbonVortex.sol

```diff
@@ -138,111 +345,505 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         }
+        return targetAmount;
+    }
 
-        // allocate rewards to caller and burn the rest
-        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function expectedTradeInput(Token token, uint128 targetAmount) public view validToken(token) returns (uint128) {
+        // revert if not enough amount available for trade
+        if (targetAmount > _amountAvailableForTrading(token)) {
+            revert InsufficientAmountForTrading();
+        }
+        Price memory currentPrice = tokenPrice(token);
+        // revert if current price is not valid
+        _validPrice(currentPrice);
+        // calculate the trade input based on the current price
+        return MathEx.mulDivF(currentPrice.sourceAmount, targetAmount, currentPrice.targetAmount).toUint128();
     }
 
     /**
-     * @dev withdraws funds held by the contract and sends them to an account
-     *
-     * requirements:
-     *
-     * - the caller must be the admin of the contract
+     * @inheritdoc ICarbonVortex
      */
-    function withdrawFunds(
+    function tokenPrice(Token token) public view returns (Price memory) {
+        // cache trading start time to save gas
+        uint32 tradingStartTime = _tradingStartTimes[token];
+        // revert if trading hasn't been enabled for a token
+        if (tradingStartTime == 0) {
+            revert TradingDisabled();
+        }
+        // get time elapsed since trading was enabled
+        uint32 timeElapsed = uint32(block.timestamp) - tradingStartTime;
+        // get initial price as set by resetTrading
+        Price memory price = _initialPrice[token];
+        // get the halflife for the token
+        uint32 currentPriceDecayHalfLife = token == _targetToken ? _targetTokenPriceDecayHalfLife : _priceDecayHalfLife;
+        // get the current price by adjusting the amount with the exp decay formula
+        price.sourceAmount = ExpDecayMath
+            .calcExpDecay(price.sourceAmount, timeElapsed, currentPriceDecayHalfLife)
+            .toUint128();
+        // return the price
+        return price;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function trade(
         Token token,
-        address payable target,
-        uint256 amount
-    ) external validAddress(target) nonReentrant onlyAdmin {
-        if (amount == 0) {
+        uint128 targetAmount
+    ) external payable nonReentrant validToken(token) greaterThanZero(targetAmount) {
+        uint128 sourceAmount;
+        if (token == _targetToken) {
+            sourceAmount = _sellFinalTargetForTarget(targetAmount);
+        } else {
+            sourceAmount = _sellTargetTokenForToken(token, targetAmount);
+        }
+        emit TokenTraded({ caller: msg.sender, token: token, sourceAmount: sourceAmount, targetAmount: targetAmount });
+    }
+
+    function _sellTargetTokenForToken(Token token, uint128 targetAmount) private returns (uint128) {
```

Comment:
the contract doesn't sell the target token for the token, it does the opposite. same for the final target function